lib/trycourier.rb

```diff
@@ -18,49 +29,75 @@ def initialize(code, message_id)
   end
 
   class Client
-    def initialize(auth_token = nil)
-      @auth_token = auth_token || ENV['COURIER_AUTH_TOKEN']
-      @uri = URI.parse('https://api.trycourier.app/send')
-      if @auth_token == nil or @auth_token == ""
-        raise InputError, "Client requires an auth_token be supplied."
+    def initialize(auth_token: nil, base_url: nil, username: nil, password: nil)
+      @base_url = if base_url
+        base_url
+      elsif ENV["COURIER_BASE_URL"]
+        ENV["COURIER_BASE_URL"]
+      else
+        "https://api.courier.com"
       end
+
+      @session = Courier::CourierAPISession.new
+
+      if auth_token
+        @session.init_token_auth(auth_token)
+      elsif ENV["COURIER_AUTH_TOKEN"]
+        @session.init_token_auth(ENV["COURIER_AUTH_TOKEN"])
+      elsif username && password
+        @session.init_basic_auth(username, password)
+      elsif ENV["COURIER_AUTH_USERNAME"] && ENV["COURIER_AUTH_PASSWORD"]
+        @session.init_basic_auth(ENV["COURIER_AUTH_USERNAME"], ENV["COURIER_AUTH_PASSWORD"])
+      end
+
+      @messages = Courier::Messages.new(@base_url, @session)
+      @profiles = Courier::Profiles.new(@base_url, @session)
+      @lists = Courier::Lists.new(@base_url, @session)
+      @events = Courier::Events.new(@base_url, @session)
+      @brands = Courier::Brands.new(@base_url, @session)
     end
 
     def send(body)
-      if not body.is_a?(Hash)
+      if !body.is_a?(Hash)
         raise InputError, "Client#send must be passed a Hash as first argument."
-      elsif body["event"] == nil
+      elsif body["event"].nil?
         raise InputError, "Must specify the 'event' key in Hash supplied to Client#send."
-      elsif body["recipient"] == nil
+      elsif body["recipient"].nil?
         raise InputError, "Must specify the 'recipient' key in Hash supplied to Client#send."
-      elsif body["data"] != nil and not body["data"].is_a?(Hash)
+      elsif !body["data"].nil? && !body["data"].is_a?(Hash)
         raise InputError, "The 'data' key in the Hash supplied to Client#send must also be a Hash."
-      elsif body["profile"] != nil and not body["profile"].is_a?(Hash)
+      elsif !body["profile"].nil? && !body["profile"].is_a?(Hash)
         raise InputError, "The 'profile' key in the Hash supplied to Client#send must also be a Hash."
       end
 
-      http = Net::HTTP.new(@uri.host, @uri.port)
-      http.use_ssl = true
-      http.verify_mode = OpenSSL::SSL::VERIFY_PEER
-
-      req = Net::HTTP::Post.new(@uri)
-      req["authorization"] = "Bearer #{@auth_token}"
-      req["content-type"] = "application/json"
-      req["User-Agent"] = "courier-ruby/#{Courier::VERSION}"
-      req.body = body.to_json
+      url = @base_url + "/send"
```

Comment:
Here, if `base_url` was part of the session, then you would only need to pass `/send` and construct it in the session.

Reply:
I do like how you cleaned this up and used a session.